### PR TITLE
Ssl context

### PR DESCRIPTION
**Pull request by mikedilger**
See #471

### CODE REVIEW

File: src/net.rs [hunk 1]
```diff
@@ -164,6 +164,11 @@ impl HttpListener {
         try!(ssl_context.set_certificate_file(cert, X509FileType::PEM).map_err(lift_ssl_error));
         try!(ssl_context.set_private_key_file(key, X509FileType::PEM).map_err(lift_ssl_error));
         ssl_context.set_verify(SSL_VERIFY_NONE, None);
+        HttpListener::https_using_context(addr, ssl_context)
+    }
+
+    /// Start listening to an address of HTTPS using the given SslContext
+    pub fn https_using_context<To: ToSocketAddrs>(addr: To, ssl_context: SslContext) -> io::Result<HttpListener> {
```
Comment (reem) [inline src/net.rs]: The convention is to use "with" for these sorts of constructors, so `https_with_context` may be appropriate.

File: src/server/mod.rs [hunk 2]
```diff
@@ -50,6 +51,13 @@ pub mod response;
 
 mod listener;
 
+#[derive(Debug)]
+enum SslConfig<'a> {
+    CertAndKey(&'a Path, &'a Path),
+    Context(SslContext),
+    None
```
Comment (reem) [inline src/server/mod.rs]: better to use `Option<SslConfig<'a>>`

File: src/server/mod.rs [hunk 3]
```diff
@@ -117,23 +134,27 @@ H: Handler + 'static,
...
-        let handler = self.handler;
+    pub fn with_listener(self, listener: L, threads: usize) -> HttpResult<Listening> {
+        with_listener(self.handler, listener, threads)
+    }
+}
 
-        debug!("threads = {:?}", threads);
-        let pool = ListenerPool::new(listener.clone());
-        let work = move |mut stream| handle_connection(&mut stream, &handler);
+fn with_listener<'a, H, L>(handler: H, mut listener: L, threads: usize) -> HttpResult<Listening>
```
Comment (reem) [inline src/server/mod.rs]: `'a` appear unused here

**Comment by mikedilger** [comment]
I've amended the commits and force-pushed the amended branch.  I think I've addressed your concerns.

**Comment by seanmonstar** [comment]
Thanks!

**Pull request merged**